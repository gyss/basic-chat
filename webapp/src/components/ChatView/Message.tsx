/** @jsx jsx */

import * as React from 'react'
import * as moment from 'moment'
import {jsx} from '@emotion/core'
import styled from '@emotion/styled'

import {IMessage} from '../../types'

export interface IProps {
  message: IMessage
  hourFormat?: '24' | '12'
  isOwnedByUser?: boolean
}

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: ${(props: Partial<IProps>) => (props.isOwnedByUser ? 'flex-end' : 'flex-start')};

  > div {
    padding: 10px;
  }
`

const Bubble = styled.div`
  position: relative;
  background-color: white;
  color: black;
  border-radius: 20px;
  min-width: 40px;
  width: fit-content;

  &::after {
    content: '';
    position: absolute;
    top: -4px;
    ${(props: Partial<IProps>) => (props.isOwnedByUser ? 'right' : 'left')}: 15px;
    width: 10px;
    height: 10px;
    background-color: white;
    transform: rotate(45deg);
  }
`

export default function Message({message, isOwnedByUser, hourFormat}: IProps) {
  return (
    <Container isOwnedByUser={isOwnedByUser}>
      <div>
        {!isOwnedByUser && message.user.username + ', '}
        {moment(message.timestamp).format(hourFormat === '24' ? 'HH:mm' : 'hh:mm a')}
      </div>
      <Bubble isOwnedByUser={isOwnedByUser}>{message.text}</Bubble>
    </Container>
  )
}
