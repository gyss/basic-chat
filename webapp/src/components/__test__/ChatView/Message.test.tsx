/** @format */

import * as React from 'react'
import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Message, {IProps} from '../../ChatView/Message'

describe('<Message>', () => {
  let props: IProps

  beforeEach(() => {
    props = {
      message: {
        id: '1',
        user: {
          id: '1',
          username: 'Pepe',
        },
        timestamp: 1594037255632,
        text: 'Hello World!',
      },
    }
  })

  test('should render', () => {
    const {asFragment} = render(<Message {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render by default in 12 hours format', () => {
    render(<Message {...props} />)

    const messageMeta = screen.queryByText('Pepe, 01:07 pm')
    expect(messageMeta).toBeInTheDocument()
  })

  test('should render in 24 hours format', () => {
    render(<Message {...props} hourFormat="24" />)

    const messageMeta = screen.queryByText('Pepe, 13:07')
    expect(messageMeta).toBeInTheDocument()
  })

  test('should render only the time when isOwnedByUser', () => {
    render(<Message {...props} isOwnedByUser />)

    const messageMeta = screen.queryByText('Pepe, 01:07 pm')
    expect(messageMeta).not.toBeInTheDocument()
  })
})
