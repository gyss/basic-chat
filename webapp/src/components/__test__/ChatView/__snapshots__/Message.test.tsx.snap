// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message> should render 1`] = `
<DocumentFragment>
  <div
    class="css-at14q"
  >
    <div>
      Pepe, 01:07 pm
    </div>
    <div
      class="css-6gx0yb"
    >
      Hello World!
    </div>
  </div>
</DocumentFragment>
`;
